solidity/core/contracts/validator-manager/OutboxValidatorManager.sol

```diff
@@ -49,25 +82,150 @@ contract OutboxValidatorManager is MultisigValidatorManager {
+     * @param _invalidProof Proof of inclusion of `_fraudulentLeaf`.
+     * @param _invalidIndex The index of the disputed leaf.
+     * @return True iff invalidity was proved.
+     */
+    function invalidCheckpoint(
+        IOutbox _outbox,
+        bytes32 _signedRoot,
+        uint256 _signedIndex,
+        bytes[] calldata _signatures,
+        bytes32 _invalidLeaf,
+        bytes32[32] calldata _invalidProof,
+        uint256 _invalidIndex
+    ) external returns (bool) {
+        require(isQuorum(_signedRoot, _signedIndex, _signatures), "!quorum");
+
+        bytes32 _invalidRoot = MerkleLib.branchRoot(
+            _invalidLeaf,
+            _invalidProof,
+            _invalidIndex
+        );
+        require(_invalidRoot == _signedRoot, "!root");
+
+        bool _invalidIfNonEmptyLeaf = _invalidIndex > _signedIndex;
+        bool _nonEmptyLeaf = _invalidLeaf != 0;
+        require(_invalidIfNonEmptyLeaf == _nonEmptyLeaf, "!invalid");
+
+        // Fail the Outbox.
+        _outbox.fail();
+        emit InvalidCheckpoint(
+            address(_outbox),
+            _signedRoot,
+            _signedIndex,
+            _signatures
+        );
+        return true;
+    }
+
+    /**
+     * @notice Determines if a quorum of validators have signed a premature checkpoint,
      * failing the Outbox if so.
-     * @dev Improper checkpoints signed by individual validators are not handled to prevent
+     * A checkpoint is premature if it commits to more messages than are present in the
+     * Outbox's merkle tree.
+     * @dev Premature checkpoints signed by individual validators are not handled to prevent
      * a single byzantine validator from failing the Outbox.
      * @param _outbox The outbox.
-     * @param _root The merkle root of the checkpoint.
-     * @param _index The index of the checkpoint.
+     * @param _signedRoot The root of the signed checkpoint.
+     * @param _signedIndex The index of the signed checkpoint.
      * @param _signatures Signatures over the checkpoint to be checked for a validator
      * quorum. Must be sorted in ascending order by signer address.
+     * @return True iff prematurity was proved.
      */
-    function improperCheckpoint(
+    function prematureCheckpoint(
         IOutbox _outbox,
-        bytes32 _root,
-        uint256 _index,
+        bytes32 _signedRoot,
+        uint256 _signedIndex,
         bytes[] calldata _signatures
-    ) external {
-        require(isQuorum(_root, _index, _signatures), "!quorum");
-        require(!_outbox.isCheckpoint(_root, _index), "!improper checkpoint");
+    ) external returns (bool) {
+        require(isQuorum(_signedRoot, _signedIndex, _signatures), "!quorum");
+        // Checkpoints are premature if the checkpoint commits to more messages
+        // than the Outbox has in its merkle tree.
+        require(_signedIndex + 1 > _outbox.count(), "!premature");
+        _outbox.fail();
+        emit PrematureCheckpoint(
```

Comment:
do you think the current `_outbox.count()` would be interesting to log?

Reply:
I do!